### PR TITLE
NeMo 2.0 In-framework deployment support

### PR DESCRIPTION
**Pull request by oyilmaz-nvidia**
# What does this PR do ?

Adds inframework deployment support and log probability output from trt-llm

### CODE REVIEW

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 1]
```diff
@@ -99,6 +103,152 @@ def to_long_tensor(self):
         return torch.tensor([self], dtype=torch.long, device='cuda')
 
 
+class MegatronLLMDeploy:
+
+    @staticmethod
+    def get_deployable(
+        nemo_checkpoint_filepath: str = None,
```
Comment (janekl) [inline nemo/deploy/nlp/megatronllm_deployable.py:110]: Typing should be `Optional[str]` here

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 2]
```diff
@@ -99,6 +103,152 @@ def to_long_tensor(self):
...
+        num_devices: int = 1,
+        num_nodes: int = 1,
+        tensor_model_parallel_size: int = 1,
```
Comment (janekl) [inline nemo/deploy/nlp/megatronllm_deployable.py:113]: Params `num_devices` & `num_nodes` are quite tied to `tensor_model_parallel_size` & `pipeline_model_parallel_size`.

How about making ` tensor_model_parallel_size` & `pipeline_model_parallel_size` optional and setting them by default to `num_devices` & `num_nodes`, respectitvely (if they are `None`)?

Some assertion for consistency like ` num_devices * num_nodes == tensor_model_parallel_size * pipeline_model_parallel_size`  would also be useful.

(Not sure how `context_parallel_size` changes all this).

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 3]
```diff
@@ -99,6 +103,152 @@ def to_long_tensor(self):
...
+        else:
+            return MegatronLLMDeployable(
+                nemo_checkpoint_filepath=nemo_checkpoint_filepath,
+                num_devices=num_devices,
+                num_nodes=num_nodes,
+            )
+
+
+class MegatronLLMDeployableNemo2(ITritonDeployable):
+    """Triton inference server compatible deploy class for a .nemo model file"""
```
Comment (janekl) [inline nemo/deploy/nlp/megatronllm_deployable.py:136]: It is no longer a `.nemo` file I think, just a folder.

File: nemo/deploy/nlp/megatronllm_deployable.py [hunk 4]
```diff
@@ -99,6 +103,152 @@ def to_long_tensor(self):
...
+            if log_probs:
+                output_log_probs = []
+                for r in results:
+                    lp = r.generated_log_probs.cpu().detach().numpy()
+                    if len(lp) == 0:
+                        output_log_probs.append([0])
+                    else:
+                        output_log_probs.append(lp)
+                output_infer["log_probs"] = np.array(output_log_probs)
+        except Exception as error:
```
Comment (janekl) [inline nemo/deploy/nlp/megatronllm_deployable.py:245]: This possibly hides errors by inserting them in output directory. Can we just raise them right away, i.e. remove the `try / catch` clause?

File: nemo/deploy/nlp/query_llm.py [hunk 5]
```diff
@@ -123,7 +119,31 @@ def query_llm(
...
+
+            log_probs_output = None
+            if "log_probs" in result_dict.keys():
+                log_probs_output = result_dict["log_probs"]
+
+            if output_type == np.bytes_:
+                if "sentences" in result_dict.keys():
+                    output = result_dict["sentences"]
+                else:
+                    return "Unknown output keyword."
```
Comment (janekl) [inline nemo/deploy/nlp/query_llm.py:132]: Should this be an error?

**Comment by janekl** [comment]
Left some minor comments, overall LGTM

**Comment by oyilmaz-nvidia** [comment]
Closing this since there is a newer version of this PR: https://github.com/NVIDIA/NeMo/pull/11523

**Pull request closed without merging**